### PR TITLE
recurse directories in DirectoryIterator

### PR DESCRIPTION
**Pull request by lunardog**
Added functionality to recurse subdirectories when using `flow_from_directory`, optionally also follow symlinks.

I find it useful since my image data is on a separate device from my code, organised in subdirectories for a reason. I think it might be useful for others too.

### CODE REVIEW

**Comment by isaacgerg** [comment]
Can you elaborate on how the class categories are derived in the recursive case?

**Comment by lunardog** [comment]
Hello @isaacgerg.

Sub-directories don't map to sub-classes in my code. The only change in the code is to traverse the subdirectories inside the class directories, rather than only pick the files in the specified directory. In case where there are no sub-directories, I believe the code works as it used to.

Imagine a scenario where the cat/dog training dataset from kaggle is supplemented with a large number of cat and dog photos from other sources:

```
/cats
/cats/cats-from-kaggle
/cats/photos-of-wild-cats
/cats/external-disk -> /mnt/external-disk/cat-photos
/cats/network-volume -> /mnt/mounted-network-volume/cat-photos
/dogs
/dogs/dogs-from-kaggle
/dogs/dog-show-archive
/dogs/external-disk -> /mnt/external-disk/dog-photos
/dogs/network-volume -> /mnt/mounted-network-volume/dog-photos
```

With subfolders and links, data sources can be added or removed without actually moving or copying files. This can be especially useful with:
 - shared network volumes, used also by other users
 - datasets with more categories than needed for the task
 - [very large collections of images](http://www.multimediacommons.org) that are chunked into subdirectories

I hope that helps.

**Comment by isaacgerg** [comment]
@lunardog This sounds good.  Just wanted to make sure I understood.

**Comment by fchollet** [comment]
Is this functionality unit-tested, by the way? If it's not, it would be good at add tests.

**Comment by lunardog** [comment]
Thanks for reminding me, @fchollet. I added a test and found a bug in the process - the iterator collected 6 images, but only 4 file names. Fixed now, should be correct.

By the way, I added some test images (public domain, from Pixabay) to `tests/integration_tests/images`. Is that a good way to go about it?

**Comment by lunardog** [comment]
The tests failed due to a timeout downloading Theano from github.

 ```
Cloning git://github.com/Theano/Theano.git to /tmp/pip-q22Mqh-build
fatal: unable to connect to github.com:
github.com[0: 192.30.253.113]: errno=Connection timed out
github.com[1: 192.30.253.112]: errno=Connection timed out
Command "git clone -q git://github.com/Theano/Theano.git /tmp/pip-q22Mqh-build" failed with error code 128 in None
The command "pip install git+git://github.com/Theano/Theano.git" failed and exited with 1 during .
Your build has been stopped.
```

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -563,14 +567,15 @@ def __init__(self, directory, image_data_generator,
...
             subpath = os.path.join(directory, subdir)
-            for fname in sorted(os.listdir(subpath)):
-                is_valid = False
-                for extension in white_list_formats:
-                    if fname.lower().endswith('.' + extension):
-                        is_valid = True
-                        break
-                if is_valid:
-                    self.nb_sample += 1
+            for root, dirs, files in os.walk(subpath, followlinks=follow_links):
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Wait, why are you changing the API from that of `os.walk`?

File: tests/integration_tests/test_image_data_tasks.py [hunk 2]
```diff
@@ -45,5 +46,16 @@ def test_image_classification():
     assert(history.history['val_acc'][-1] > 0.85)
 
 
+@keras_test
+def test_directory_iterator():
+    generator = ImageDataGenerator()
+    directory_iterator = generator.flow_from_directory('./tests/integration_tests/images')
```
Comment (fchollet) [inline tests/integration_tests/test_image_data_tasks.py]: Do not include images in the git tree. If you need images, generate them. We have tools for that in `preprocessing.image`.

**Comment by fchollet** [comment]
Rather than adding an integration test, please add a unit test for `flow_from_directory` in `tests/keras/processing/test_image.py`.

File: keras/preprocessing/image.py (hunk 1 above)
Comment (lunardog) [inline keras/preprocessing/image.py]: what do you mean?

File: tests/integration_tests/test_image_data_tasks.py (hunk 2 above)
Comment (lunardog) [inline tests/integration_tests/test_image_data_tasks.py]: Will fix. Thank you.

File: keras/preprocessing/image.py (hunk 1 above)
Comment (lunardog) [inline keras/preprocessing/image.py]: if you mean the `followlinks`, `follow_links` difference, I wanted to match the underscore convention in Keras. I realize it collides with `os.walk`.

File: keras/preprocessing/image.py [hunk 3]
```diff
@@ -195,8 +195,9 @@ def load_img(path, grayscale=False, target_size=None):
 
 
 def list_pictures(directory, ext='jpg|jpeg|bmp|png'):
```
Comment (lunardog) [inline keras/preprocessing/image.py]: Is this used anywhere?

**Comment by lunardog** [comment]
@fchollet 
1. I put the test in processing, not integration_Tests
2. I added sorting
3. I kept the underscore in `follow_links` to match Keras API

Sorry about all the `oops` commits. Should I squash and make a new PR?

**Comment by lunardog** [comment]
Made a [new, properly squashed PR](https://github.com/fchollet/keras/pull/4637). Closing.

**Pull request closed without merging**